Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -1839,37 +1861,172 @@ void CGContextSetPatternPhase(CGContextRef context, CGSize phase) {
+                                       [&state, glyphRun](CGContextRef context, ID2D1DeviceContext* deviceContext) {
+                                           deviceContext->DrawGlyphRun(D2D1::Point2F(0, 0),
+                                                                       glyphRun,
+                                                                       state.fillBrush.Get(),
+                                                                       DWRITE_MEASURING_MODE_NATURAL);
+                                           return S_OK;
+                                       });
+        RETURN_IF_FAILED(hr);
+    } else {
+        // Using device space here gives us finer-grained control over the transform.
+        HRESULT hr = DrawToCommandList(
+            _kCGCoordinateModeDeviceSpace,
+            nullptr,
+            &textCommandList,
+            [&state, &glyphRun, &deviceTransform, &textTransform](CGContextRef context, ID2D1DeviceContext* deviceContext) {
+                CGAffineTransform runningGlobalTransform = deviceTransform;
+                // Text scaling and rotation apply to each glyph relative to its origin, so we must draw each glyph transformed
+                // independently
+                DWRITE_GLYPH_RUN individualGlyphRun{ glyphRun->fontFace,
+                                                     glyphRun->fontEmSize,
+                                                     1, // Since this is glyph by glyph, glyphCount is one
+                                                     glyphRun->glyphIndices,
+                                                     nullptr,
+                                                     nullptr,
+                                                     glyphRun->isSideways,
+                                                     glyphRun->bidiLevel };
+                D2D1_POINT_2F origin{ 0, 0 };
+                // Iterate through every glyph by incrementing pointer in glyphIndices array
+                for (uint32_t i = 0; i < glyphRun->glyphCount; ++i, ++(individualGlyphRun.glyphIndices)) {
+                    CGAffineTransform finalTextTransform = CGAffineTransformConcat(textTransform, runningGlobalTransform);
+                    deviceContext->SetTransform(__CGAffineTransformToD2D_F(finalTextTransform));
+                    deviceContext->DrawGlyphRun(origin, &individualGlyphRun, state.fillBrush.Get(), DWRITE_MEASURING_MODE_NATURAL);
+
+                    // Uses glyphAdvances to move each glyph
+                    runningGlobalTransform = CGAffineTransformTranslate(runningGlobalTransform, glyphRun->glyphAdvances[i], 0);
+                }
+                return S_OK;
+            });
+        RETURN_IF_FAILED(hr);
+    }
+
+    RETURN_IF_FAILED(DrawImage(textCommandList.Get()));
+
+    ClearPath();
+    return S_OK;
+}
+
+// Internal: used by CoreText.
+void CGContextDrawGlyphRun(CGContextRef context, const DWRITE_GLYPH_RUN* glyphRun) {
```

Comment:
i would suggest to add it at least in the actual function (makes sense to validate it, rather than expecting the user to give it in)
In the future someone decide to be less cautious and call it, we would end up with a crash.